### PR TITLE
Block directories in theme assets

### PR DESCRIPTION
**Pull request by amcaplan**
### WHY are these changes introduced?

This is the followup to #1881. We were failing in an unexpected way because we assumed users would not have subdirectories in their asset directories, rather than enforcing this behavior. @macournoyer [graciously confirmed](https://github.com/Shopify/shopify-cli/pull/1881#issuecomment-1005870905) that asset subdirectories are not supported. So we're going to fail with a proper validation message.

### WHAT is this pull request doing?

Before: [Bugsnag](https://app.bugsnag.com/shopify/shopify-cli/errors/615e274eccef1c0007f99247?filters%5Bevent.since%5D=30d)

Screen Shot 2022-01-16 at 21 48 22

After:

Screen Shot 2022-01-16 at 21 49 19

Note the small additional benefit that following this PR, we're validating earlier in the process, preventing the frustration of syncing to 100% and only then discovering things are broken.

### How to test your changes?

Create a theme with a subdirectory within `assets/` and try to `shopify theme serve`

### Update checklist

- [ ] I've added a CHANGELOG entry for this PR (if the change is public-facing)
- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows).
- [x] I've left the version number as is (we'll handle incrementing this when releasing).

### CODE REVIEW

**Review by macournoyer — APPROVED** [review]
That's awesome! So much better error message now. Thanks Ariel 🙏

**Review by macournoyer — APPROVED** [review]
_EDIT: whoops, github hiccup... sorry for the double post_

File: lib/shopify_cli/theme/theme.rb [hunk 1]
```diff
@@ -35,8 +35,16 @@ def json_files
...
+        paths = root.glob(pattern)
+        if raise_on_dir
+          paths.each do |path|
+            if ::File.directory?(path)
+              @ctx.abort(@ctx.message('theme.serve.error.invalid_subdirectory', path.to_s))
+            end
+          end
+        end
+        paths.map { |path| File.new(path, root) }
       end
```
[the remark this answers is not in the document]
Comment (amcaplan) [reply lib/shopify_cli/theme/theme.rb:48]: Love it, thanks for the suggestion @gonzaloriestra!

File: lib/project_types/theme/messages/messages.rb [hunk 2]
```diff
@@ -114,6 +114,10 @@ module Messages
           error: {
             address_binding_error: "Couldn't bind to localhost."\
               " To serve your theme, set a different address with {{command:%s theme serve --host=<address>}}",
+            invalid_subdirectory: "The directory %s is in an invalid location."\
+              " Nonstandard subdirectories are not supported within themes."\
+              " For more details on supported theme directory structure, see"\
+              " https://shopify.dev/themes/architecture#directory-structure-and-component-types."
```
[the remark this answers is not in the document]
Comment (MeredithCastile) [reply lib/project_types/theme/messages/messages.rb:120]: @amcaplan What's the task the user is trying to do when they'd hit this message? Adding assets to the subdirectory, right?

First draft pending your answer: "Assets can't be added to subdirectories. Add them directly to your assets folder (/assets)."

[the remark this answers is not in the document]
Comment (amcaplan) [reply lib/project_types/theme/messages/messages.rb:120]: @MeredithCastile have a look at the PR message images, it happens when attempting to serve the theme. So they wouldn't necessarily have just added this file.

There's also the general point that while we see this in assets mostly (presumably users want to have more organization around their assets), technically we could see failures in other areas as well, if the user has violated the officially supported directory structure.

[the remark this answers is not in the document]
Comment (MeredithCastile) [reply lib/project_types/theme/messages/messages.rb:120]: @amcaplan Ah, I understand now. Thanks for talking me through this earlier. Let's do this:

image

> Your directory structure isn't supported.
> 
> Move any files to a parent folder, then delete unsupported subdirectories.
> 
> Required directory structure: https://shopify.dev/themes/architecture#directory-structure-and-component-types

Nice to have would be the indentation and bullet before the doc / link (as in the Figma screenshot).

[the remark this answers is not in the document]
Comment (amcaplan) [reply lib/project_types/theme/messages/messages.rb:120]: Slight adjustment so they know what to delete:

Screen Shot 2022-03-22 at 20 42 00

**Pull request merged**